File: java/convert_sorted_list_to_binary_search_tree.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode sortedListToBST(ListNode head) {
        if (head==null) return null;
        ListNode one=head, prev=null, next, two=head;
        while (two!=null && two.next!=null) {
            prev=one;
            one=one.next;            
            two=two.next.next;            
        }
        
        TreeNode node=new TreeNode(one.val);
        if (prev!=null) prev.next=null;
        next=one.next;
        one.next=null;               
        node.left=sortedListToBST(one==head?null:head);
        node.right=sortedListToBST(next);
        return node;
    }
}

File: cpp/convert_sorted_list_to_binary_search_tree.cpp
109. Convert Sorted List to Binary Search Tree  QuestionEditorial Solution  My Submissions
Total Accepted: 76583 Total Submissions: 244580 Difficulty: Medium
Given a singly linked list where elements are sorted in ascending order, convert it to a height balanced BST.

Hide Company Tags Zenefits
Hide Tags Depth-first Search Linked List
Hide Similar Problems (M) Convert Sorted Array to Binary Search Tree

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <string>
#include <vector>
#include <queue>
#include <stack>
#include <unordered_set>
#include <map>

class Solution {
public:
    TreeNode *sortedListToBST(ListNode *head) {
    	if (head == NULL) return NULL;

    	ListNode *first=head, *second=head->next, *prev=NULL;
    	while (second!=NULL)
    	{
    		prev = first;
    		first = first->next;
    		second = second->next;
    		if(second!=NULL) second=second->next;
    	}

    	TreeNode *res = new TreeNode(first->val);
    	ListNode *before;
    	ListNode *after = first->next;

    	first->next=NULL;
    	if (prev!=NULL) { prev->next=NULL; before = head; }
    	else { before = NULL; }
    	res->left = sortedListToBST(before);
    	res->right = sortedListToBST(after);

    	return res;
    }
};

2.
class Solution {
public:
    TreeNode* sortedListToBST(ListNode* head) {
        if (head==NULL) return NULL;
        ListNode *one=head, *two=head, *prev=NULL, *next;
        while (two && two->next) {
            prev=one;
            one=one->next;
            two=two->next->next;
        }
        
        TreeNode *n=new TreeNode(one->val);
        if (prev) prev->next=NULL;
        next=one->next;
        one->next=NULL;
        n->left=sortedListToBST(head!=one?head:NULL);
        n->right=sortedListToBST(next);
        return n;
    }
};

int main()
{
	return 0;
}

